src/util/log/fd_log.c

```diff
@@ -628,6 +631,82 @@ fd_log_private_2( int          level,
   abort();
 }
 
+char const * 
+fd_log_private_hexdump_msg ( char const * descr,
+                             void const * blob,
+                             ulong        sz ) {
+  if( FD_UNLIKELY( !blob ) ) {
+    FD_LOG_WARNING(( "Hexdump NULL blob ptr" ));
+    return NULL;
+  }
+
+  if( FD_UNLIKELY( !descr ) ) {
+    FD_LOG_WARNING(( "Hexdump NULL blob description" ));
+    return NULL;
+  }
+
+  if( FD_UNLIKELY( strlen( descr )==0 ) ) {
+    FD_LOG_WARNING(( "Hexdump empty blob description" ));
+    return NULL;
+  }
+
+  if( FD_UNLIKELY( strlen( descr )>FD_LOG_HEXDUMP_BLOB_DESCRIPTION_MAX_LEN ) ) {
+    FD_LOG_WARNING(( "Hexdump blob description too long" ));
+    return NULL;
+  }
+
+  ulong blob_sz = sz;
+  int is_truncated = 0;
+  if( FD_UNLIKELY ( blob_sz>FD_LOG_HEXDUMP_MAX_INPUT_BLOB_SZ ) ) {
+    blob_sz = FD_LOG_HEXDUMP_MAX_INPUT_BLOB_SZ;
+    is_truncated = 1;
+  }
+  
+  char * log_buf_ptr = log_msg;   /* used by FD_LOG_HEXDUMP_ADD_TO_LOG_BUF macro */
+  int num_bytes_written = 0;      /* used by the FD_LOG_HEXDUMP_ADD_TO_LOG_BUF macro. signed because *printf() return 'int'. */
+
+  /* Blob description for easier log grepping.
+     Just print 'msg: empty' for a zero length buffer and return,
+     and make it clear when truncation happened if that is the case. */
+  if( FD_LIKELY( blob_sz ) && !is_truncated ) {
+    FD_LOG_HEXDUMP_ADD_TO_LOG_BUF( "HEXDUMP %s:\n", descr );
+  } else if( FD_UNLIKELY( is_truncated ) ) {
+    FD_LOG_HEXDUMP_ADD_TO_LOG_BUF( "HEXDUMP %s (truncated):\n", descr );
+  } else {
+    FD_LOG_HEXDUMP_ADD_TO_LOG_BUF( "HEXDUMP %s: empty\n", descr );
+    return log_msg;
+  }
```

Comment:
Recommend including the size explicitly here for easy of readability (and to help log readers understand how much info was lost in the case of a truncated hexdump).